### PR TITLE
[GPU] Several optimization to reduce runtime host overhead

### PR DESCRIPTION
**Pull request by yeonbok**
### Details:
 - Reduced unnecessary barrier & set arguments
 - Calculate flushing timing only considering gpu impls

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp [hunk 1]
```diff
@@ -280,9 +280,9 @@ struct typed_primitive_impl_ocl : public typed_primitive_impl<PType> {
                                    << (needs_completion_event ? " has_completion_event=true" : "") << std::endl;
 
             auto ev = stream.enqueue_kernel(*_kernels[kd_idx], params, args, tmp_events, needs_completion_event);
-            new_events.push_back(ev);
+            if (_kernel_data.needs_sub_kernels_sync)
+                new_events.push_back(ev);
             all_events.push_back(ev);
-
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:285]: Looks like it may break pipeline for events synchronization type
Let's assume that current Concat has N kernels and depends on N input events
Before:
1) Concat_0 kernel uses `tmp_events(events)` as dependency events, adds its' output event to `new_events` and saves it in `tmp_events` 
2) Concat_1 kernel uses `tmp_events` (which includes Concat_0's kernel dependency, which depends on 3 input events), so Concat_1 have to be executed only after Concat_0 (and 3 original input dependency respectively)
...

After: 
1) Concat_0 kernel uses `tmp_events(events)` as dependency events, but doesn't add its output event to `new_events` (because of needs_sub_kernels_sync) and copies empty vector to `tmp_events` 
2) Concat_1 kernel uses empty `tmp_events` vector and can be executed w/o any dependencies at all

So, in case of `needs_sub_kernels_sync=false` we need to pass original dependencies to `stream.enqueue_kernel()` for the second and next sub-kernels

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 2]
```diff
@@ -1034,10 +1034,12 @@ void network::execute_impl(const std::vector<event::ptr>& events) {
         }
 
         execute_primitive(inst, events);
-        executed_prims++;
+        if (!inst->get_node().get_selected_impl()->is_cpu())
+            executed_prims++;
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/network.cpp:1038]: Does it really improve host overheads? I'm curious, because during tuning this flush period I tried the same approach, and it showed worse results than unconditional increment. I don't have detailed comparison related to counting of gpu impls only vs any impl, but IIRC it was visible on bert-small-uncased-whole-word-masking-squad-0001-fp16 model

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:285]: Thanks for your comment, I agree your point, but the reason why it is still okay with as is, the barrier is already injected in the first kernel. But for safety I will update the depdencies. :)

File: src/plugins/intel_gpu/src/graph/network.cpp (hunk 2 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/network.cpp:1038]: Thanks for sharing your experience. Logically I cannot understand why cpu impl should affect the flushing efficiency... Let me check with bert-small-uncased model.

Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/network.cpp:1038]: But I decided not to investigate for this more, because this part did not show the major perf impact. => To remove

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp (hunk 1 above)
Comment (sshlyapn) [reply src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp]: Yeah, for barriers sync it was completely okay. My point was mainly about `sync_methods::events` synchronization type, but now it looks good, thank you! :)

**Review by sshlyapn — APPROVED** [review]

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 3]
```diff
@@ -1034,10 +1034,10 @@ void network::execute_impl(const std::vector<event::ptr>& events) {
         }
 
         execute_primitive(inst, events);
-        executed_prims++;
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/src/graph/network.cpp:1037]: Could you please revert this change as well?

Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/network.cpp:1037]: oops reverted!
